Clown Abominations Part 1 [Clown Rat]

ItsDzordz opened the pull request:
![image](https://github.com/user-attachments/assets/d6d95c4f-60f6-4305-ab1e-20300a043afd)
# Description

This PR adds The Clown Rat, the incarnation of honk mothers mischef. The rat will split into 3-4 rats upon death from blunt trauma (Up to 3 generations of rats, each generation has less hp and is visually smaller) , the rats can be incnrated to prevent spliting (They die faster due to heat dmg). The final generation of rats can be gibbed for a bike horn implant, or skinned for 3 cream pies. it honks when you squeeze or pet it, its blood is slippery. it has a small chance to spawn in the vent critter event. They can be ghost controlled, and can drag objects althou very slowley. [Mob is non hostile]

Sindicate has took the honk mother creation and twisted it into a weapon of mass chaos.
For 10 tc (discounted 6) agents can buy a dehydrated Clown Crazed Rat Box. The first generation of rats behave the same as a Dehydrated carp. Further generations will attack and bite anything in sight. The rat will now split due to all physical trauma so heat is recomanded to deal with the threat. [Mob IS hostile and part of the sindicate faction]

---

# TODO

- [x] Add Clown rat
- [x] Add Hostile syndicate version of the Clown rat [Crazed Clown Rat]
- [x] Add the Crazed Clown Rat to uplink
- [x] Add Clown Rat to critter vent spawn event
- [ ] Make the Clown rat Slippery (Optional, but very much wanted)
- [ ] Balancing with Mainteners

---


<details><summary><h1>Media</h1></summary>
<p>

![image](https://github.com/user-attachments/assets/1ad1c0d6-f445-4c3e-8aaf-6fcf89e33ce1)
![image](https://github.com/user-attachments/assets/59ede6f8-36f6-4553-b738-d2a1618e5d29)
![image](https://github.com/user-attachments/assets/39c97404-c573-480b-bf0c-2bccbf0ba30d)
![image](https://github.com/user-attachments/assets/34952fd9-3a52-417a-8fac-4cf18d68e895)
![image](https://github.com/user-attachments/assets/2ef1a1c9-f063-4c44-9dcc-e833f51fedb2)
![image](https://github.com/user-attachments/assets/a001822e-468a-474b-9ce1-f9da00c7a331)


</p>
</details>

---

# Changelog
:cl:
- add: Clown Rat and its Crazed variant

ItsDzordz committed:
Clown rat is born

All the malice of the trickery of the honk mother made manifest

ItsDzordz committed:
Clown Rat for the Syndicate

Added a syndicate version of the clown rat, to the uplink.
Added the regular clown rat to the vent spawn event.

ItsDzordz committed:
Bug fix

Fixed the critter event error

ItsDzordz committed:
Bug fix #2

Correcting errors from bot checks

ItsDzordz committed:
Bug fix 3

Removing non existent damge type

Mnemotechnician commented:
Just gonna mention this, be careful when spawning many entities at once. Players with low network bandwidths or low entity budgets (configured in client-side network settings) can experience freezes when many entities simultaneously enter their PVS (and in case of mobs entity counts quickly stack up because every mob has organs, actions, solution entities, and more)

ItsDzordz commented:
yeah im tweaking values to try to make it a bit less lag heavy, do you know how i can remove the gibblets from the current methods of death for these rats?

ItsDzordz committed:
Value tweak

ItsDzordz commented:
Can i have any info on what is beeing discussed about this PR?

Memeji commented:
@FoxxoTrystan @Mnemotechnician Your input on the above?

Memeji reviewed (changes requested):
I'm fine with this addition.
The test fails are related though, and will need to be resolved.

M3739 commented on Resources/Locale/en-US/_Floof/store/uplink-catalog.ftl:
```diff
@@ -11,3 +11,6 @@ uplink-EnergySabre-desc = Crafted by Fae, the CyberSun Weapons Designer, a brill
 uplink-typewriter-bundle-desc = An inconspicuous briefcase that contains a Typewriter SMG, three magazines and an outfit fit for any aspiring mafiosos.
+
+uplink-dehydrated-clown-rat-name = Dehydrated Crazed Clown Rat Box
+uplink-dehydrated-clown-rat-desc = A box containing a modified clown rat, that after hydration seeks chaos and enemies of the syndicate. Upon exparation due to physical trauma it will split into 3-4 more and those will split into even more reaching the final generation of rats. DISCLAMER remeber to pet the rats head for the first generation to be friendly the further generation are too rampid to  ...
```
```suggestion
uplink-dehydrated-clown-rat-desc = A box containing a modified clown rat, that after hydration, seeks chaos and enemies of the syndicate. Upon expiration due to physical trauma, it will split into 3-4 more and those will split further, eventually reaching the final generation of rats. 
    DISCLAIMER: remember to pet the rats head for the first generation to be friendly, as further generations are too rabid to discern between you and foe.
```

Note: The new line and spaces is to attempt to make the "DISCLAIMER:" start on a new line, similar to the holoparasite description in [`uplink-catalog.ftl` at line 106](https://github.com/M3739/FS-Yesod/blob/fc592af2975b7db64772e966c127e28388d3b002/Resources/Locale/en-US/store/uplink-catalog.ftl#L106) to make it stand out a bit more.

ItsDzordz replied:
looks good will commit it

M3739 commented on Resources/Prototypes/_Floof/Entities/Mobs/NPCs/clown_abominations.yml:
```diff
@@ -0,0 +1,324 @@
+  name: dehydrated crazed clown rat box
+  parent: DehydratedSpaceCarp
+  id: DehydratedClownRatBox
+  description: The malice of the honk mother weponised by the syndicate in a compact package. Just pet its noggin and add water!
```
```suggestion
  description: The malice of the honkmother, weaponized by the syndicate in a compact package. Just pet its noggin and add water!
```
After a quick lookup, it turns out whether or not "honkmother" is a proper noun or common noun is... inconsistent. But it is always spelled as "honkmother" instead of "honk mother". 

This can be observed as "Honkmother" in  [`instruments_misc.yml` at line 156](https://github.com/Fansana/floofstation1/blob/fc592af2975b7db64772e966c127e28388d3b002/Resources/Prototypes/Entities/Objects/Fun/Instruments/instruments_misc.yml#L156), and "honkmother" in [`ghost-role-component.ftl` at line 218](https://github.com/Fansana/floofstation1/blob/fc592af2975b7db64772e966c127e28388d3b002/Resources/Locale/en-US/ghost/roles/ghost-role-component.ftl#L218).

ItsDzordz replied:
Good to know thak you for spell checking it i ussly do it as the last thing after all the bugs are ironed out, so i do appriate it. will commit the change

M3739 commented on Resources/Prototypes/GameRules/events.yml:
```diff
@@ -264,12 +264,15 @@
     - id: SpawnPointGhostRatKing
       prob: 0.001
 
+
```
Random new line?

ItsDzordz replied:
oh yeah ill have to remove that, i first wanted to add the code about the chance of the clown rat spawning below, but i had to move it up for consitency and and soem issues, seems i left a line by accdent

ItsDzordz replied:
adjusted

M3739 commented on Resources/Prototypes/_Floof/Entities/Mobs/NPCs/clown_abominations.yml:
```diff
@@ -0,0 +1,324 @@
+    rootTask:
+      task: SimpleHostileCompound
+  - type: Destructible
+    thresholds:
+    - trigger:
+        !type:DamageTypeTrigger
+        damageType: Slash
+        damage: 40
+      behaviors:
+      - !type:SpawnEntitiesBehavior
+        spawnInContainer: true
+        spawn:
+          MobCrazedClownRatFinalHatchling:
+            min: 3
+            max: 4
+      - !type:GibBehavior { }
+    - trigger:
+        !type:DamageTypeTrigger
+        damageType: Blunt
+        damage: 40
+      behaviors:
+      - !type:SpawnEntitiesBehavior
+        spawnInContainer: true
+        spawn:
+          MobCrazedClownRatFinalHatchling:
+            min: 3
+            max: 4
+      - !type:GibBehavior { }
+    - trigger:
+        !type:DamageTypeTrigger
+        damageType: Piercing
+        damage: 40
+      behaviors:
+      - !type:SpawnEntitiesBehavior
+        spawnInContainer: true
+        spawn:
+          MobCrazedClownRatFinalHatchling:
+            min: 3
+            max: 4
+      - !type:GibBehavior { }
```
After a quick test in a development environment, it would appear that `MobCrazedClownRatHatchling` is not inheriting the heat ashing mechanic that `MobClownRatHatchling` would have: 
![image](https://github.com/user-attachments/assets/bee6bdab-8e16-4219-aaf3-735af9237acb)

ItsDzordz replied:
I did some testing and turns out (from how i understand this), when inhereting from a parent and having the same componenet ("- type: Destructible") but diffrent parammiters the one on the child will completley override the parent even if the child is not  modifing these certain paramiters (individualt triggers in the componenet). since they MobCrazedClownRatFinalHatchling has no destructible component it behaves completley like the parent in terms of perishing. compared to MobCrazedClownRatHatchling. maybe not most elegant solution but i reapplied the ashing due to heat damage triggers mannualy and it fixed the issue

ItsDzordz replied:
![image](https://github.com/user-attachments/assets/31f01968-bf79-42f4-a66d-31786ab13c26)
![image](https://github.com/user-attachments/assets/c13a65f3-cf8e-4244-8377-bdd86fd525df)

ItsDzordz replied:
Next commit will contain this fix

M3739 commented on Resources/Prototypes/_Floof/Entities/Mobs/NPCs/clown_abominations.yml:
```diff
@@ -0,0 +1,324 @@
+    rootTask:
+      task: SimpleHostileCompound
+  - type: Destructible
+    thresholds:
+    - trigger:
+        !type:DamageTypeTrigger
+        damageType: Slash
+        damage: 60
+      behaviors:
+      - !type:SpawnEntitiesBehavior
+        spawnInContainer: true
+        spawn:
+          MobCrazedClownRatHatchling:
+            min: 3
+            max: 4
+      - !type:GibBehavior { }
+    - trigger:
+        !type:DamageTypeTrigger
+        damageType: Blunt
+        damage: 60
+      behaviors:
+      - !type:SpawnEntitiesBehavior
+        spawnInContainer: true
+        spawn:
+          MobCrazedClownRatHatchling:
+            min: 3
+            max: 4
+      - !type:GibBehavior { }
+    - trigger:
+        !type:DamageTypeTrigger
+        damageType: Piercing
+        damage: 60
+      behaviors:
+      - !type:SpawnEntitiesBehavior
+        spawnInContainer: true
+        spawn:
+          MobCrazedClownRatHatchling:
+            min: 3
+            max: 4
+      - !type:GibBehavior { }
```
Same as `MobCrazedClownRatHatchling`, it would appear that this mob is also not inheriting the heat ashing mechanic of `MobClownRat`: 
![image](https://github.com/user-attachments/assets/24373cf5-e4ef-47a7-9436-ae8a283689aa)

ItsDzordz replied:
will resolve this coverstation and will keep up the issue in the post above since its the same issue about inheretance

M3739 reviewed (changes requested):
Done a proofread through both the uplink description and the box description.

The uplink description also got a slight rewrite to enhance clarity.

However, there are perceived bugs that I think need to be addressed, primarily with how the parenting of the clown rat prototypes are being done.

Allow me to recite a conversation I had with @Mnemotechnician (tagging in case I get anything wrong) when I encountered this same issue when I was working on #561 dealing with the `interfaces` component and it's `interface` complex field:

> M3739: _"As it turns out, specifying a component in a prototype will completely overwrite the same component being inherited from the parent prototype?"_

> Mnemotechnician: _"No, specifying a component in the child will not overwrite it. It will still be inherited from the first parent that specifies it before being modified in the child. However, if you overwrite a complex field inside a component (a nested DataDefinition, a list, a dictionary, or anything else), that particular field will not be inherited and will be fully overwriten unless it has the AlwaysPushInheritance attribute"_

> M3739: _"Oh, so specifying  the `interfaces` and it's following BUI keys will do it, since `interface` is one of those complex fields?"_

> Mnemotechnician: _"yeah"_
_"it will override that list with an empty one because it doesn't have the AlwaysPushInheritance attribute"_

For reference, the `MobCrazedClownRatFinalHatchling` seems to inherit the `destructible` component properly, as it ashes when damaged with heat, which seems to be proper behavior.

But because you specify the `destructible` component with the `trigger` complex field, it would seem to be fully overwriting the component in the respective prototypes.

ItsDzordz committed:
Update Resources/Locale/en-US/_Floof/store/uplink-catalog.ftl

Co-authored-by: M3739 <47579354+M3739@users.noreply.github.com>

ItsDzordz committed:
Update Resources/Prototypes/_Floof/Entities/Mobs/NPCs/clown_abominations.yml

Co-authored-by: M3739 <47579354+M3739@users.noreply.github.com>

ItsDzordz committed:
burn inhertance bug fix

ItsDzordz committed:
Update events.yml

ItsDzordz commented:
so what does the test say is wrong? i dont understand them
> I'm fine with this addition. The test fails are related though, and will need to be resolved.

M3739 commented:
> so what does the test say is wrong? i dont understand them
> 
> > I'm fine with this addition. The test fails are related though, and will need to be resolved.

![image](https://github.com/user-attachments/assets/fd7a6aa4-e4d0-4fb6-a643-6a3370b0025a)

Apparently, it is saying that it gains in total value after being  "destroyed". As in sell price. It seems like it does not like both the ash spawning and the hydra effect of killing the rats.

You may want to consider options such as the `StaticPrice` component in order to set a fixed price on entity prototypes.

ItsDzordz commented:
Well when i think about the spliting rat rasing in the stock market i die a little inside. so im going to set a static price on the main parent will see if that is enough

ItsDzordz committed:
Price fix

ItsDzordz committed:
Price fix 2

ItsDzordz committed:
ash amount fix

ItsDzordz commented:
ok reducing amout of ash didnt solve the issue as well as settign the main parent price to 0, only other thing i could think off is to somehow remove the ghost role property (and i really want people to be able to control this mob) or figure out how to remove gibblets from the gibbing of the rat. Problem, i dont really know how to achive that, any ideas for a fix? ill manually set the price of all forms to 0, and let the test run. im going to take a look tommrow if it solved the issue since is late, if all tests run without any issues feel free to merge.

ItsDzordz committed:
price fix 3

ItsDzordz commented:
Well nothing changed, how does this calculate the value?
![image](https://github.com/user-attachments/assets/a5fd1e2f-69be-481b-a6e7-8484da514364)

ItsDzordz committed:
Price fix 5

ItsDzordz commented:
with further testing i came to a conclusion that this test will always resolve i na failure, becaouse of the muliplactive nature of the mob, i ask that this test be skipped. Since this does not affect gamplay nor perfomence

M3739 commented:
I managed to find the test itself where it happens [`MaterialArbitrageTest.cs`](https://github.com/Fansana/floofstation1/blob/6a7ad5cb215b2a256b339051d2de6b425ca01e41/Content.IntegrationTests/Tests/MaterialArbitrageTest.cs#L141-L143),  and a comment on calculating the things to test:
```
// This test just blindly assumes that ALL spawn entity behaviors get triggered. In reality, some entities
// might only trigger a subset. If that starts being a problem, this test either needs fixing or needs to
// get an ignored prototypes list.
```
And a comment on the price test:
```
// Check cargo sell price
// several constructible entities have no sell price
// also this test only really matters if the entity is also purchaseable.... eh..
```

Considering that it's only buyable via TC, and not purchasable via cargo, I believe we can exercise the executive decision of skipping the test for now.

ItsDzordz committed:
Price fix 6

ItsDzordz committed:
Test re-run

teruning test since it seems the current error is irrelevent to this pr

M3739 reviewed (approved):
LGTM!

ItsDzordz committed:
Faction fix

added the overwritten faction from the parnets so they are attacked by shiva and laika

Memeji reviewed (approved):
Tests pass, looks good to me

ItsDzordz commented:
Did some testing and they fight other mobs like they should now, its good to go
